clippy_lints/src/methods/iter_cloned_collect.rs

```diff
@@ -1,16 +1,21 @@
 use crate::methods::utils::derefs_to_slice;
 use clippy_utils::diagnostics::span_lint_and_sugg;
-use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::ty::{get_iterator_item_ty, is_type_diagnostic_item};
 use rustc_errors::Applicability;
 use rustc_hir as hir;
 use rustc_lint::LateContext;
+use rustc_middle::ty;
 use rustc_span::sym;
 
 use super::ITER_CLONED_COLLECT;
 
 pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, method_name: &str, expr: &hir::Expr<'_>, recv: &'tcx hir::Expr<'_>) {
-    if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(expr), sym::Vec)
+    let expr_ty = cx.typeck_results().expr_ty(expr);
+    if is_type_diagnostic_item(cx, expr_ty, sym::Vec)
         && let Some(slice) = derefs_to_slice(cx, recv, cx.typeck_results().expr_ty(recv))
+        && let ty::Adt(_, args) = expr_ty.kind()
+        && let Some(iter_item_ty) = get_iterator_item_ty(cx, cx.typeck_results().expr_ty(recv))
+        && iter_item_ty.peel_refs() == args.type_at(0)
```

Comment:
You may end up peeling too many levels. For example, the following will be a false negative:

```rust
    let a = [&&String::new()];
    let v: Vec<&&String> = a.iter().cloned().collect();
```

I would have went with something like the following:

```suggestion
        && let ty::Ref(_, ty, _) = iter_item_ty.kind()
        && *ty == args.type_at(0)
```

Reply:
Good point. Updated